extensions/amp-mega-menu/0.1/amp-mega-menu.js

```diff
@@ -144,47 +120,53 @@ export class AmpMegaMenu extends AMP.BaseElement {
       AmpEvents.DOM_UPDATE,
       this.domUpdateHandler_
     );
-    this.documentElement_.removeEventListener(
-      'click',
-      this.documentClickHandler_
-    );
-    this.documentElement_.removeEventListener(
-      'keydown',
-      this.documentKeyDownHandler_
-    );
-    this.unregisterMenuItemListeners_();
     // Ensure that menu is closed when hidden via media query.
     this.collapse_();
     return true;
   }
 
+  /**
+   * Create a new mask element and append it to the header if present,
+   * otherwise to the component itself.
+   * @return {!Element} the mask that was created.
+   * @private
+   */
+  createMaskElement_() {
+    const mask = this.document_.createElement('div');
+    mask.classList.add('i-amphtml-mega-menu-mask');
+    mask.setAttribute('aria-hidden', 'true');
+    // append mask to header so that all children of header appear above the mask
+    const maskParent =
+      closestAncestorElementBySelector(this.element, 'header') || this.element;
+    maskParent.classList.add('i-amphtml-mega-menu-mask-parent');
+    maskParent.appendChild(mask);
+    return mask;
+  }
+
   /**
    * Find all menu items under this mega menu and add appropriate classes,
    * attributes and event listeners to its children.
    * @private
    */
   registerMenuItems_() {
-    // remove existing event listeners and re-add them at the end.
-    this.unregisterMenuItemListeners_();
-    this.items_ = toArray(
-      scopedQuerySelectorAll(this.element, 'nav > * > li')
-    ).filter(item => {
-      // skip if the item has already been registered.
+    this.items_ = [];
+    const items = toArray(scopedQuerySelectorAll(this.element, 'nav > * > li'));
+    items.forEach(item => {
+      // check if the item has already been registered.
       if (item.classList.contains('i-amphtml-mega-menu-item')) {
-        return true;
+        this.items_.push(item);
+        return;
       }
       const heading =
         scopedQuerySelector(item, '> button') ||
         scopedQuerySelector(item, '> [role=button]');
       const content = scopedQuerySelector(item, '> [role=dialog]');
+      // register item only if heading element is present.
```

Comment:
I think when @kristoferbaxter suggested use of a filter function, he meant use the filter function to filter down to only the items in the array you want to operate on. For example, you could also write this as: 
```
this.items_ = toArray(scopedQuerySelectorAll(this.element, 'nav > * > li'));
this.items_.filter(item => !item.classList.contains('i-amphtml-mega-menu-item'))
.foreach(item => {
  const heading = ...
  ...
  this.registerMenuItem_(item, heading, content);
});
```

Reply:
hmm the thing is I only want `this.items_` to include items with heading elements; so i can filter out the elements without headings first, but then in the `foreach` I'll need to get a reference to `heading` again to pass to `registerMenuItem_`.

Reply:
I've made some changes that should allow me to use the filter and foreach pattern now. For any `<li>` that conforms to the specs laid out in #25210, we'll register it as a menu item and include it in `this.items_`. The only case not covered by validation but should be disallowed is when there're two elements under an item and neither is a button/role=button, in which case I've added a `userAssert` to warn against this. Hopefully the documentation will make it clear.